treewide: replace wpad-basic-wolfssl default

neheb opened the pull request:
The newly merged mbedtls backend is smaller and has fewer ABI related issues than the wolfSSL one.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
treewide: replace wpad-basic-wolfssl default

The newly merged mbedtls backend is smaller and has fewer ABI related
issues than the wolfSSL one.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

aparcar commented:
LGTM, I'd merge this

chbgdn commented:
I use wpad-basic-mbedtls along month on MT7986A (Redmi AX6000), not have problems at all. Merge it please :)

rozhuk-im reviewed (commented):
LGTM

rozhuk-im commented:
There is interesting PR with wolfssl: https://github.com/openwrt/openwrt/pull/10765

neheb commented:
ping @Ansuel

motolav commented:
You're missing ipq807x

neheb commented:
Fixed

motolav commented:
One other issue, ipq807x includes libwolfsslcpu-crypto by default so it should be removed for consistency

ventod commented:
`libwolfssl` ( `make menuconfig -> Libraries -> SSL` ) is still selected, shouldn't this be not compiled now? `libustream-wolfssl` ( `make menuconfig -> Libraries` ) is selected instead of `libustream-mbedtls` ( `make menuconfig -> Libraries` ). Am I missing something? This I merge this locally wrong?

Also does this take advantage of kmod-ltq-deu-vr9 ( #4326 ) anyhow ?

(I have a BT Home Hub 5A)

neheb commented:
That's libustream. I'll make a separate commit for that.

neheb committed:
treewide: replace libustream-wolfssl with -mbedtls

Previous commit does the same for wpad-basic. Also matches
DEFAULT_VARIANT in ustream-ssl Makefile.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb commented:
Done.

neheb commented:
unrelated CI failure

Ansuel commented:
@aparcar @hauke looking at the meeting thing we should start testing this widely so we can actually consider replacing wolfssl to this?

I would merge this as is and disaster handling it?

neheb commented:
rebased. again,

@Ansuel what disaster?

ventod commented:
Testing this on my BTHH5a. What should I look for (errors/functionalities/...)?

neheb commented:
it works perfectly fine. Only potentially untested configuration is some exotic WPA enterprise setup.

hauke commented:
I think we should remove `libwolfsslcpu-crypto` from ipq807x too.

We do not have `uencrypt` for mbedtls, but I do not think this is blocking. 

This looks good.

aparcar commented:
@hauke as discussed during the [last meeting](https://openwrt.org/meetings/20230124) we decided to have mbedtls on SMALL_FLASH devices and switch to openssl for the rest. How do you want to proceed here?

Borromini commented:
@aparcar I read those same meeting notes and I am not seeing that decision anywhere? What would be the rationale behind that call? Are there still essential features mbedTLS cannot provide for hostapd and the likes e.g.?

IMO it only brings more overhead:
* Providing both OpenSSL and mbedTLS based images depending on flash size adds to maintenance burden.
* More point releases for security issues in the SSL libraries (since OpenSSL isn't default for any target AFAIK only WolfSSL vulnerabilities require a new point release right now).
* Many devices not considered SMALL_FLASH will be pushed into that segment more quickly than needed by burdening them with OpenSSL. *All* those image builds will break at some point, and need to be moved manually to SMALL_FLASH.

I can understand if you'd move certain big flash targets (x86, ARM64) to OpenSSL by default, but the first two points would still apply. And you'll need an extra BIG_FLASH setting to help with that.

aparcar commented:
> @aparcar I read those same meeting notes and I am not seeing that decision anywhere?

That part wasn't documented, sorry! I could be remembering this wrong, maybe @hauke @pepe2k @ynezz @wigyori @lynxis @dangowrt @adschm remember?

> What would be the rationale behind that call? 

Based on @ynezz OpenSSL is generally faster and more mature at this point. MbedTLS is smaller (even WolfSSL) and therefore a better fit for SMALL_FLASH. WolfSSL caused many ABI issues so we wanted to look for an alternative. Maybe @jow- or @nbd168 can elaborate on this point.

> Are there still essential features mbedTLS cannot provide for hostapd and the likes e.g.?

Based on @dangowrt testing it's a full replacement for regular use cases. I don't know the status of special cases.

> * Providing both OpenSSL and mbedTLS based images depending on flash size adds to maintenance burden.

True, but hopefully in a more predictable way than with WolfSSL

> * More point releases for security issues in the SSL libraries (since OpenSSL isn't default for any target AFAIK only WolfSSL vulnerabilities require a new point release right now).

Being the developer of this attended sysupgrade thing I don't fear point releases but I agree, more tagging and release notes.

> * Many devices not considered SMALL_FLASH will be pushed into that segment more quickly than needed by burdening them with OpenSSL. _All_ those image builds will break at some point, and need to be moved manually to SMALL_FLASH.

I think that's a question of your definition. We could say that WPA3 (and thereby the dependence on crypto libs) is pushing more devices into SMALL_FLASH etc, right?

> I can understand if you'd move certain big flash targets (x86, ARM64) to OpenSSL by default, but the first two points would still apply. And you'll need an extra BIG_FLASH setting to help with that.

That could be another way.

dhewg commented:
Adding another point of view:

Quoting the OpenSSL website:
`The previous LTS version (the 1.1.1 series) is also available and is supported until 11th September 2023`

And v1.1.1 is what's in tree. Shipping 23.xx with v1.1.1 per default should be a no-go. A v3.0 bump hence becomes mandatory. But there isn't even a v3.0 PR is sight. Even if that appears, that'll effectively changing two SSL provider in a single release, just before branching.

aparcar commented:
@dhewg I agree, I think @hauke meant to ask about anyone working on v3.0, I'm sure he does in a day or two. Sounds like a switch to mbedtls by default and a performance target like x86 with OpenSSL v3.0 later on could be the best option.

neheb commented:
I’m against defaulting to OpenSSL. Nice in theory and potential speed (has anyone tested?) but more CVEs come from there. There’s the 1.1 and 3.0 issue too.

aparcar commented:
@neheb can you please take care of the comment of @hauke ? I think then this is fine to move on

neheb commented:
I don’t think it’s related to this PR.

aparcar commented:
> I don’t think it’s related to this PR.

I think @cotequeiroz will look into adding mbedtls support but could you add another commit removing libwolfsslcpu-crypto please?

neheb committed:
ipq807x: remove libwolfsslcpu-crypto dependency

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb commented:
Done.

aparcar commented:
Merged, thanks.

msdos03 commented:
Excuse me, I didn't find any options enabling crypto HW acceleration in mbedtls. Marvell kirkwood, mvebu target's have a HW crypto engine. I didn't find anything says that mbedtls supports AF_ALG or Cryptodev acceleration. If mbedtls didn't support that, that's a problem to me and every targets with a HW crypto engine.

dangowrt commented:
@msdos03 You are free to still use wolfssl or openssl if needed on devices offering enough space for that (both significantly larger). What was changed here is merely the *default* implementation. Also note that wifi encryption itself is **not** carried out by a userspace library, it is used only for the initial handshake and key exchange in WPA3 as well as things like EAP-TLS (WPA-Enterprise). Hence you should not see any notable difference in wifi performance in a typical setup serving up to a few dozens of clients.

msdos03 commented:
@dangowrt Luci-ssl uses crypto more often, but there is no luci-ssl using wolfssl now. Why not select wolfssl by default on targets with HW crypto engine?

neheb commented:
is marvell's mv_crypto even useful? on my system I see:

```
(afalg) AFALG engine support
 [AES-128-CBC, AES-192-CBC, AES-256-CBC]
     [ available ]
```

all of which is useless as everyone abandoned CBC because of BEAST and other such attacks.

msdos03 commented:
`
name         : hmac(sha1)
driver       : mv-hmac-sha1
module       : kernel
priority     : 300
refcnt       : 1
selftest     : passed
internal     : no
type         : ahash
async        : yes
blocksize    : 64
digestsize   : 20

name         : hmac(md5)
driver       : mv-hmac-md5
module       : kernel
priority     : 300
refcnt       : 1
selftest     : passed
internal     : no
type         : ahash
async        : yes
blocksize    : 64
digestsize   : 16

name         : sha1
driver       : mv-sha1
module       : kernel
priority     : 300
refcnt       : 1
selftest     : passed
internal     : no
type         : ahash
async        : yes
blocksize    : 64
digestsize   : 20

name         : md5
driver       : mv-md5
module       : kernel
priority     : 300
refcnt       : 1
selftest     : passed
internal     : no
type         : ahash
async        : yes
blocksize    : 64
digestsize   : 16

name         : cbc(aes)
driver       : mv-cbc-aes
module       : kernel
priority     : 300
refcnt       : 1
selftest     : passed
internal     : no
type         : skcipher
async        : yes
blocksize    : 16
min keysize  : 16
max keysize  : 32
ivsize       : 16
chunksize    : 16
walksize     : 16

name         : ecb(aes)
driver       : mv-ecb-aes
module       : kernel
priority     : 300
refcnt       : 1
selftest     : passed
internal     : no
type         : skcipher
async        : yes
blocksize    : 16
min keysize  : 16
max keysize  : 32
ivsize       : 0
chunksize    : 16
walksize     : 16

name         : cbc(des3_ede)
driver       : mv-cbc-des3-ede
module       : kernel
priority     : 300
refcnt       : 1
selftest     : passed
internal     : no
type         : skcipher
async        : yes
blocksize    : 8
min keysize  : 24
max keysize  : 24
ivsize       : 8
chunksize    : 8
walksize     : 8

name         : ecb(des3_ede)
driver       : mv-ecb-des3-ede
module       : kernel
priority     : 300
refcnt       : 1
selftest     : passed
internal     : no
type         : skcipher
async        : yes
blocksize    : 8
min keysize  : 24
max keysize  : 24
ivsize       : 0
chunksize    : 8
walksize     : 8

name         : cbc(des)
driver       : mv-cbc-des
module       : kernel
priority     : 300
refcnt       : 1
selftest     : passed
internal     : no
type         : skcipher
async        : yes
blocksize    : 8
min keysize  : 8
max keysize  : 8
ivsize       : 8
chunksize    : 8
walksize     : 8

name         : ecb(des)
driver       : mv-ecb-des
module       : kernel
priority     : 300
refcnt       : 1
selftest     : passed
internal     : no
type         : skcipher
async        : yes
blocksize    : 8
min keysize  : 8
max keysize  : 8
ivsize       : 0
chunksize    : 8
walksize     : 8
`
I have these ones.

neheb commented:
All are useless. ECB is a basic and totally insecure mode used nowhere. CBC was used everywhere back in the day. After BEAST and similar attacks, everyone abandoned it for CTR mode.

The only possible application of mv_crypto in 2023 is accelerating disk encryption. Of course you have to use the older CBC-ESSIV mode.

@cotequeiroz probably knows more.

msdos03 commented:
@neheb Are MD5 and SHA1 useless too?

neheb commented:
Yes. HW acceleration for those is for bulk hashing, not one off.

msdos03 commented:
@neheb Anyway, add back luci wolfssl is good. Why not do it?

neheb commented:
This PR is for defaults, nothing else. And no wolfssl is not good. Since the conversion, OpenWrt has had to deal with constant ABI breakage.

msdos03 commented:
@neheb Yes, that is in another repo. They just removed luci wolfssl and add luci mbedtls.
